### PR TITLE
Stdlib: Add function composition combinator

### PR DESCRIPTION
**Pull request by laelath**
Add `Fun.compose` and `Stdlib.(@.)` for function composition. Useful when wanting to apply multiple functions when wanting to map over a list as `List.map (f @. g) l` exposes the high-level logic in a clearer manner than `List.map (fun x -> f (g x)) l`. The choice of `@.` was made because Haskell uses `.` for function composition and there is probably a clearer choice that could be made for OCaml.

### CODE REVIEW

**Comment by nojb** [comment]
My opinion based on a quick look:
- In favour of adding `Fun.compose`,
- Not really in favour of adding the infix operator: 1) I think the general direction is not to add anything else to `Stdlib` itself to avoid polluting the global namespace, and 2) infix operators tend to harm code readability, and I'm not sure we want to encourage that.

**Comment by laelath** [comment]
While overuse of infix operators can certainly harm code readability, I think that more sparing/deliberate usages of them can make things clearer. For the purposes of this PR consider `f @. g @. h` vs `compose f (compose g h)`. Maybe the infix operators could be added without polluting the global namespace by defining a `Fun.Operators` sub module that can be `open`ed to access?

**Comment by Octachron** [comment]
There is also the issue that the choice of `@.` is non-standard. A [quick look on sherlocode](('\[a-z\] -> '\[a-z\]) -> ('\[a-z\] -> '\[a-z\]) -> '\[a-z\] -> '\[a-z\]) yields the following operator name for `compose`. 
- `|.`, `-|`
- `<<<`, `<<`
- `$`, `%`
- `@*`
- `++`

In other words, the name of the operator is controversial, and it is better to split this discussion on the operator from the simple `compose` function.

I am personally in favor of settling a name for the composition operator, but this is better done in a large discussion rather than in a PR.

**Comment by dbuenzli** [comment]
This has already been discussed to death [here](https://github.com/ocaml/ocaml/pull/2097).

**Comment by xvw** [comment]
In Preface we use `f % g` for `f . g` (and `f %> g` for `g . f`, wich is never really used)  and we have some case where it is useful and probably more readable than `(fun x -> f (g x))`. And since we have a module Arrow for `type ('a, 'b) t = 'a -> 'b)` we can also use `>>>` and `<<<` but we take the habit to use `%`.

**Comment by laelath** [comment]
I removed the infix operator for now

**Review by nojb — APPROVED** [review]
LGTM

I am in favour of this addition. As it touches the standard library, a second approval from a core dev is needed.

**Comment by nojb** [comment]
Friendly ping. This is a simple PR to review; we still need a second reviewer; volunteers welcome!

**Comment by dbuenzli** [comment]
I looked at it today. It's ok with me if someone wants to ack on my behalf.

**Review by yallop — APPROVED** [review]
Approved (on behalf of @dbuenzli, but I'm happy with it, too)

**Comment by nojb** [comment]
Thanks for the prompt review! I took the liberty of amending the Changes file accordingly.

**Pull request merged**
